tests_integ/helpers/base_test.py

```diff
@@ -0,0 +1,379 @@
+import logging
+import os
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+from unittest.case import TestCase
+
+import boto3
+import pytest
+import yaml
+from botocore.exceptions import ClientError
+from botocore.config import Config
+from tests_integ.helpers.deployer.deployer import Deployer
+from tests_integ.helpers.helpers import transform_template, verify_stack_resources, generate_suffix, create_bucket
+from tests_integ.helpers.file_resources import FILE_TO_S3_URI_MAP, CODE_KEY_TO_FILE_MAP
+
+LOG = logging.getLogger(__name__)
+STACK_NAME_PREFIX = "sam-integ-stack-"
+S3_BUCKET_PREFIX = "sam-integ-bucket-"
+
+
+class BaseTest(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.tests_integ_dir = Path(__file__).resolve().parents[1]
+        cls.resources_dir = Path(cls.tests_integ_dir, "resources")
+        cls.template_dir = Path(cls.resources_dir, "templates", "single")
+        cls.output_dir = Path(cls.tests_integ_dir, "tmp")
+        cls.expected_dir = Path(cls.resources_dir, "expected", "single")
+        cls.code_dir = Path(cls.resources_dir, "code")
+        cls.s3_bucket_name = S3_BUCKET_PREFIX + generate_suffix()
+        cls.session = boto3.session.Session()
+        cls.my_region = cls.session.region_name
+        cls.s3_client = boto3.client("s3")
+        cls.api_client = boto3.client("apigateway")
+        cls.lambda_client = boto3.client("lambda")
+        cls.iam_client = boto3.client("iam")
+        cls.api_v2_client = boto3.client("apigatewayv2")
+        cls.sfn_client = boto3.client("stepfunctions")
+
+        cls.file_to_s3_uri_map = FILE_TO_S3_URI_MAP
+        cls.code_key_to_file = CODE_KEY_TO_FILE_MAP
+
+        if not cls.output_dir.exists():
+            os.mkdir(str(cls.output_dir))
+
+        cls._upload_resources(FILE_TO_S3_URI_MAP)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._clean_bucket()
+
+    @classmethod
+    def _clean_bucket(cls):
+        """
+        Empties and deletes the bucket used for the tests
+        """
+        s3 = boto3.resource("s3")
```

Comment:
Why use resource here but client everywhere else? Seems like resource is a much better way to handle things in boto3 in general?

Reply:
It seems `resource` is more high level that `client`, we'll check to see if we can use resource everywhere.
We needed the `resource` here to the files list in a paginated way.

Reply:
```client``` is low level and has typically maps 1:1 with the AWS service API while ```resource``` is high level but does not provide 100% API coverage of AWS services. Most of the case we use ```client```, here we use ```resource``` since it provides a easier way to lists all the objects compared to the ```client```.